sequencer/src/auction.rs

```diff
@@ -0,0 +1,103 @@
+use crate::{state::FeeAmount, NamespaceId, Transaction};
+use hotshot_types::data::ViewNumber;
+use std::num::NonZeroU64;
+
+/// State Machine for Auction
+#[derive(Clone, Copy)]
+struct Auction {
+    /// current phase.
+    phase: AuctionPhase,
+    /// A structure to find which Phase should be the current one.
+    // This could probably be an enum with some restructuring.
+    possible_phases: [AuctionPhase; 3],
+}
+
+impl Auction {
+    fn update(mut self, view: ViewNumber) {
+        self.phase = self
+            .possible_phases
+            .into_iter()
+            .find(|phase| (view > phase.start && view < phase.end))
+            .unwrap();
+    }
+    fn recv_bid(&self, view: ViewNumber, bid: BidTx) {
+        self.update(view);
+        match self.phase.kind {
+            AuctionPhaseKind::Bid => self.send_txn(bid.as_txn()),
+            _ => unimplemented!(),
+        }
+    }
+    fn send_txn(&self, txn: Transaction) {
+        unimplemented!();
+    }
+}
+
+// - needs to be configured in genesis block
+// - needs to be updatable
+/// Configuration for the auction system
+struct AuctionConfig {
+    bid_phase_num_views: NonZeroU64,
+    auction_phase_num_views: NonZeroU64,
+    sequencing_phase_num_views: NonZeroU64,
+}
+
+/// Uniquely identifies an auction for sequencing rights of namespaces in the network
+#[derive(Clone, Copy)]
+struct AuctionId(u64);
+
+/// Uniquely identifies one auction phase for a specific auction
+#[derive(Clone, Copy)]
+struct AuctionPhaseId(AuctionId, AuctionPhaseKind);
+
+/// Describes one auction phase for a specific auction
+#[derive(Clone, Copy)]
+struct AuctionPhase {
+    id: AuctionPhaseId,
+    kind: AuctionPhaseKind,
+    start: ViewNumber,
```

Comment:
Reading over some discussion it looks like timing will be determined by block instead of view, so these might need to be updated to be `BlockHeight`.

Reply:
This has since been clarfied